File: bindings/lua/insight/complex.lua
--- Complex number operations.
--
-- Provides functions for creating, querying, and manipulating
-- complex-valued arrays.
--
-- @module insight.complex

local native = require("_insight")
local M = {}

local function _wrap(names, fn)
  return function(...)
    if select("#", ...) == 1 and type(select(1, ...)) == "table" then
      local t = select(1, ...)
      local has_names = false
      for k, _ in pairs(t) do
        if type(k) ~= "number" then
          has_names = true
          break
        end
      end
      if has_names then
        local pos = {}
        for i, name in ipairs(names) do
          pos[i] = t[name]
          if pos[i] == nil then
            pos[i] = t[i]
          end
        end
        return fn(table.unpack(pos, 1, #names))
      end
    end
    return fn(...)
  end
end

--- Check if an array has a complex dtype.
-- @tparam Array x Input array.
-- @bool True if the array has a complex dtype (complex64 or complex128).
M.is_complex = _wrap({ "x" }, function(x)
  return native.is_complex(x)
end)

--- Check if an array uses legacy complex storage (last dim = 2).
-- @tparam Array x Input array.
-- @bool True if the array uses the legacy real-imag layout.
M.has_complex_shape = _wrap({ "x" }, function(x)
  return native.has_complex_shape(x)
end)

--- Convert a real array to complex (imaginary part = 0).
-- @tparam Array real Real-valued input array.
-- @tparam[opt] Array imag Imaginary part. If nil, zeros are used.
-- @treturn Array Complex-valued array.
M.to_complex = _wrap({ "real", "imag" }, function(real, imag)
  if imag then
    return native.to_complex(real, imag)
  end
  return native.to_complex(real)
end)

--- View a real array with last dim=2 as complex (zero-copy).
-- @tparam Array x Real input array whose last dimension is 2.
-- @treturn Array Complex view.
M.as_complex = _wrap({ "x" }, function(x)
  return native.as_complex(x)
end)

--- View a complex array as real with last dim=2 (zero-copy).
-- @tparam Array x Complex input array.
-- @treturn Array Real view with doubled last dimension.
M.as_real = _wrap({ "x" }, function(x)
  return native.as_real(x)
end)

--- Extract the real part of a complex array.
-- @tparam Array z Complex input array.
-- @treturn Array Real-valued array.
M.real = _wrap({ "z" }, function(z)
  return native.real(z)
end)

--- Extract the imaginary part of a complex array.
-- @tparam Array z Complex input array.
-- @treturn Array Real-valued array.
M.imag = _wrap({ "z" }, function(z)
  return native.imag(z)
end)

return M

File: bindings/python/insight/complex.py
"""Complex number operations.

Functions:
    is_complex, has_complex_shape, to_complex, as_complex
    as_real, real, imag
"""

from __future__ import annotations

from insight._insight import Array  # noqa: F401
from insight._insight import (
    is_complex as _native_is_complex,
    has_complex_shape as _native_has_complex_shape,
    to_complex as _native_to_complex,
    as_complex as _native_as_complex,
    as_real as _native_as_real,
    real as _native_real,
    imag as _native_imag,
)


def is_complex(x: "Array") -> bool:
    """Check whether an array has a complex data type.

    Args:
        x: Input array.

    Returns:
        True if the array's dtype is complex64 or complex128.
    """
    return _native_is_complex(x)


def has_complex_shape(x: "Array") -> bool:
    """Check whether an array uses legacy complex storage format.

    The legacy format stores real and imaginary parts as the last
    dimension of size 2, using a real-valued dtype (e.g., float32).

    Args:
        x: Input array.

    Returns:
        True if the last dimension exists and has size 2.
    """
    return _native_has_complex_shape(x)


def to_complex(real: "Array", imag: "Array | None" = None) -> "Array":
    """Convert real-valued array(s) to a complex array.

    If only ``real`` is provided, the imaginary part is set to zero.
    If both ``real`` and ``imag`` are provided, the result is
    ``real + imag * j``.

    Args:
        real: Real part array.
        imag: Imaginary part array (optional). If provided, must have
            the same shape as ``real``.

    Returns:
        Complex array with shape ``(*real.shape, 2)`` (interleaved
        real/imag storage).
    """
    if imag is None:
        return _native_to_complex(real)
    return _native_to_complex(real, imag)


def as_complex(x: "Array") -> "Array":
    """View a real array as a complex array (zero-copy).

    The input must have its last dimension equal to 2 (interleaved
    real and imaginary parts). The dtype is promoted from F32 to C32
    or from F64 to C64.

    Args:
        x: Real array with shape ``(..., 2)``.

    Returns:
        Complex array view with shape ``(...)``.
    """
    return _native_as_complex(x)


def as_real(x: "Array") -> "Array":
    """View a complex array as a real array (zero-copy).

    The output gains an extra last dimension of size 2, containing the
    interleaved real and imaginary parts.

    Args:
        x: Complex array (dtype C32 or C64).

    Returns:
        Real array view with shape ``(*x.shape, 2)``.
    """
    return _native_as_real(x)


def real(z: "Array") -> "Array":
    """Extract the real part of a complex array.

    Args:
        z: Complex array with shape ``(..., 2)``.

    Returns:
        Real array view with the last dimension removed, containing
        only the real parts.
    """
    return _native_real(z)


def imag(z: "Array") -> "Array":
    """Extract the imaginary part of a complex array.

    Args:
        z: Complex array with shape ``(..., 2)``.

    Returns:
        Real array view with the last dimension removed, containing
        only the imaginary parts.
    """
    return _native_imag(z)


__all__ = [
    "is_complex",
    "has_complex_shape",
    "to_complex",
    "as_complex",
    "as_real",
    "real",
    "imag",
]
